db/migrations/postgres/000106_threadmemberships_cleanup.up.sql

```diff
@@ -0,0 +1,15 @@
+DELETE FROM threadmemberships WHERE EXISTS (
```

Comment:
@agnivade you're correct, changing the order seems better :)

```sql
mattermost_test=# EXPLAIN ANALYZE DELETE FROM threadmemberships WHERE EXISTS (
mattermost_test(#     SELECT
mattermost_test(#         threadmemberships.*
mattermost_test(#     FROM
mattermost_test(#         threadmemberships
mattermost_test(#         LEFT JOIN channelmembers ON channelmembers.userid = threadmemberships.userid
mattermost_test(#         JOIN threads ON threads.PostId = threadmemberships.postid
mattermost_test(#             AND threads.channelid = channelmembers.channelid
mattermost_test(#         JOIN channels ON channels.id = threads.channelid
mattermost_test(#     WHERE
mattermost_test(#         threads.postid = threadmemberships.postid
mattermost_test(#         AND channelmembers.channelid IS NULL
mattermost_test(#         AND (Channels.Type != 'D'
mattermost_test(#             OR Channels.Type != 'G')
mattermost_test(# );
                                                                             QUERY PLAN

---------------------------------------------------------------------------------------------------------------------------
------------------------------------------
 Delete on threadmemberships  (cost=17.04..31.24 rows=420 width=6) (actual time=0.012..0.013 rows=0 loops=1)
   InitPlan 1 (returns $3)
     ->  Nested Loop  (cost=0.58..17.04 rows=1 width=0) (actual time=0.009..0.010 rows=0 loops=1)
           ->  Nested Loop  (cost=0.44..16.71 rows=1 width=140) (actual time=0.009..0.010 rows=0 loops=1)
                 Join Filter: ((channelmembers.userid)::text = (threadmemberships_1.userid)::text)
                 ->  Nested Loop  (cost=0.29..16.34 rows=1 width=280) (actual time=0.009..0.009 rows=0 loops=1)
                       ->  Index Only Scan using channelmembers_pkey on channelmembers  (cost=0.14..8.16 rows=1 width=140)
(actual time=0.008..0.009 rows=0 loops=1)
                             Index Cond: (channelid IS NULL)
                             Heap Fetches: 0
                       ->  Index Scan using idx_threads_channel_id_last_reply_at on threads  (cost=0.15..8.17 rows=1 width=
140) (never executed)
                             Index Cond: ((channelid)::text = (channelmembers.channelid)::text)
                 ->  Index Only Scan using threadmemberships_pkey on threadmemberships threadmemberships_1  (cost=0.15..0.3
5 rows=2 width=140) (never executed)
                       Index Cond: (postid = (threads.postid)::text)
                       Heap Fetches: 0
           ->  Index Scan using channels_pkey on channels  (cost=0.14..0.33 rows=1 width=70) (never executed)
                 Index Cond: ((id)::text = (threads.channelid)::text)
                 Filter: ((type <> 'D'::channel_type) OR (type <> 'G'::channel_type))
   ->  Result  (cost=0.00..14.20 rows=420 width=6) (actual time=0.011..0.011 rows=0 loops=1)
         One-Time Filter: $3
         ->  Seq Scan on threadmemberships  (cost=0.00..14.20 rows=420 width=6) (never executed)
 Planning time: 2.410 ms
 Execution time: 0.148 ms
(22 rows)

Reply:
That's good but you have to _force_ the join order in the query. Otherwise Postgres is free to select any order it wishes to. See this slide: https://docs.google.com/presentation/d/1lmE1dYyrUVphFYcrCkRza_Etm6rJD7uLH-sDw-tyEF8/edit#slide=id.g13c2f642a31_0_16

Set `join_collapse_limit` to 1, and verify that different orders do indeed result in different plans.

Reply:
@agnivade it seems that we cannot join threadmemberships, with channelmembers without joining threads first, we need that ChannelId.